src/backend/src/heap_state/btc_user_pending_tx_state.rs

```diff
@@ -0,0 +1,423 @@
+use candid::Principal;
+use ic_cdk::api::management_canister::bitcoin::Utxo;
+use std::collections::HashMap;
+
+#[allow(dead_code)]
+const MAX_PENDING_TRANSACTIONS: usize = 1000;
+#[allow(dead_code)]
+const DAY_IN_NS: u64 = 24 * 60 * 60 * 1_000_000_000;
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct StoredPendingTransaction {
+    txid: Vec<u8>,
+    utxos: Vec<Utxo>,
+    created_at_timestamp_ns: u64,
+}
+
+type BitcoinAddress = String;
+type PendingTransactionsMap = HashMap<BitcoinAddress, Vec<StoredPendingTransaction>>;
+
+#[allow(dead_code)]
+pub struct BtcUserPendingTransactions {
+    /// Map of `user_principal` to `PendingTransactionsMap`;
+    pending_transactions_map: HashMap<Principal, PendingTransactionsMap>,
+    /// Maximum number of transactions stored per principal.
+    max_pending_transactions: usize,
+}
+
+impl BtcUserPendingTransactions {
+    #[allow(dead_code)]
+    pub fn new(max_pending_txs: Option<usize>) -> Self {
+        Self {
+            pending_transactions_map: HashMap::new(),
+            max_pending_transactions: max_pending_txs.unwrap_or(MAX_PENDING_TRANSACTIONS),
+        }
+    }
+
+    /// Returns the pending transactions of a specific principal per address.
```

Comment:
Can multiple users share the same Bitcoin address?  Not if the Bitcoin address is derived from the principal, I guess, but maybe in future.  My concern is that if one user makes a transaction, it may not appear for others sharing the same address.  This is not going to happen if we commit to not sharing addresses.

Reply:
At the moment, we don't share addreses.

This would be a problem not only for sharing, but if multiple users can make transactions from the same address. Which at the moment is not supported in CFS either.

But it's something to keep in mind for the future. I'll add a comment.